sources/engine/Stride.Physics/Collision.cs

```diff
@@ -67,70 +28,56 @@ internal void Destroy()
 
-        public ChannelMicroThreadAwaiter<ContactPoint> ContactUpdate()
+        internal Collision(PhysicsComponent a, PhysicsComponent b)
         {
-            if (destroyed)
-                throw new InvalidOperationException("The collision object has been destroyed.");
-
-            return ContactUpdateChannel.Receive();
+            ColliderA = a;
+            ColliderB = b;
         }
 
-        internal Channel<ContactPoint> ContactEndedChannel;
-
-        public ChannelMicroThreadAwaiter<ContactPoint> ContactEnded()
+        /// <summary>
+        /// The returned collection contains the previous contacts, new contacts are under <see cref="Contacts"/>
+        /// </summary>
+        public ChannelMicroThreadAwaiter<HashSet<ContactPoint>> ContactChanged()
         {
-            if (destroyed)
-                throw new InvalidOperationException("The collision object has been destroyed.");
-
-            return ContactEndedChannel.Receive();
+            return ColliderA.Simulation.ContactChanged(this);
         }
 
         public async Task Ended()
         {
-            if (destroyed)
+            if (ColliderA.Simulation.CurrentCollisions.Contains(this) == false)
                 throw new InvalidOperationException("The collision object has been destroyed.");
 
             Collision endCollision;
             do
             {
                 endCollision = await ColliderA.CollisionEnded();
-            }
-            while (!endCollision.Equals(this));
+            } while (!endCollision.Equals(this));
         }
 
         public override bool Equals(object obj)
         {
-            var other = (Collision)obj;
-            return other != null && ((other.ColliderA == ColliderA && other.ColliderB == ColliderB) || (other.ColliderB == ColliderA && other.ColliderA == ColliderB));
+            return obj is Collision other && Equals(other);
         }
 
-        public override int GetHashCode()
+        public bool Equals(Collision other)
         {
-            unchecked
-            {
-                var result = ColliderA?.GetHashCode() ?? 0;
-                result = (result * 397) ^ (ColliderB?.GetHashCode() ?? 0);
-                return result;
-            }
+            return (Equals(ColliderA, other.ColliderA) && Equals(ColliderB, other.ColliderB)) 
+                   || (Equals(ColliderB, other.ColliderA) && Equals(ColliderA, other.ColliderB));
         }
 
-        internal bool InternalEquals(PhysicsComponent a, PhysicsComponent b)
+        public override int GetHashCode()
         {
-            return (ColliderA == a && ColliderB == b) || (ColliderB == a && ColliderA == b);
+            int aH = ColliderA.GetHashCode();
+            int bH = ColliderB.GetHashCode();
+            // This ensures that a pair of components will return the same hash regardless
+            // of if they are setup as [b,a] or [a,b]
+            return aH > bH ? HashCode.Combine(aH, bH) : HashCode.Combine(bH, aH);
```

Comment:
Are we sure `aH != bH`?
I suppose `ColliderA` and `ColliderB` can't be the same?
Still, there could be a collision hash (especially if poorly implemented).

Reply:
I don't think I understand what you meant here ?
If they have the same hash values the order you combine them in doesn't matter;
Let's say ``aH`` is ``10``, now if ``aH == bH`` then ``bH`` must be ``10`` as well, so the call would look like ``HashCode.Combine(10, 10)``, swapping the tens around wont change the resulting hash generated since it's the same value ... ?

Reply:
Yes obviously it's fine, please disregard my comment, I must have been too tired when I reviewed :)